src/intrinsic/simd.rs

```diff
@@ -156,6 +160,163 @@ pub fn generic_simd_intrinsic<'a, 'gcc, 'tcx>(
+        let byte_vector_type_size = type_size_bytes.max(16);
+        let byte_vector_type = bx.context.new_vector_type(bx.u8_type, type_size_bytes);
+        let long_byte_vector_type = bx.context.new_vector_type(bx.u8_type, byte_vector_type_size);
+
+        let hi_nibble_elements: Vec<_> = (0u8..16)
+            .map(|x| bx.gcc_int(bx.u8_type, x.reverse_bits() as _))
+            .chain((16..byte_vector_type_size).map(|_| bx.gcc_zero(bx.u8_type)))
+            .collect();
+        let hi_nibble = bx.context.new_rvalue_from_vector(None, long_byte_vector_type, &hi_nibble_elements);
+
+        let lo_nibble_elements: Vec<_> = (0u8..16)
+            .map(|x| bx.gcc_int(bx.u8_type, (x.reverse_bits() >> 4) as _))
+            .chain((16..byte_vector_type_size).map(|_| bx.gcc_zero(bx.u8_type)))
+            .collect();
+        let lo_nibble = bx.context.new_rvalue_from_vector(None, long_byte_vector_type, &lo_nibble_elements);
+
+        let mask = bx.context.new_rvalue_from_vector(
+            None,
+            long_byte_vector_type,
+            &vec![bx.gcc_int(bx.u8_type, 0x0f); byte_vector_type_size as _]);
+
+        let swapped = simd_bswap(bx, args[0].immediate());
+        let byte_vector = bx.context.new_bitcast(None, swapped, byte_vector_type);
+
+        // We're going to need to extend the vector with zeros to make sure that the types are the
+        // same, since that's what new_rvalue_vector_perm expects.
+        let byte_vector = if byte_vector_type_size > type_size_bytes {
+            let mut byte_vector_elements = Vec::with_capacity(byte_vector_type_size as _);
+            for i in 0..type_size_bytes {
+                let idx = bx.gcc_int(bx.u32_type, i as _);
+                let val = bx.extract_element(byte_vector, idx);
+                byte_vector_elements.push(val);
+            }
+            for _ in type_size_bytes..byte_vector_type_size {
+                let val = bx.gcc_zero(bx.u8_type);
+                byte_vector_elements.push(val);
+            }
+            bx.context.new_rvalue_from_vector(None, long_byte_vector_type, &byte_vector_elements)
+        } else {
+            bx.context.new_bitcast(None, byte_vector, long_byte_vector_type)
+        };
+
+        let four = bx.gcc_int(bx.u8_type, 4);
+        let four_vec = bx.context.new_rvalue_from_vector(
+            None,
+            long_byte_vector_type,
+            &vec![four; byte_vector_type_size as _]);
+        let masked_hi = bx.gcc_lshr(byte_vector, four_vec) & mask;
+        let masked_lo = byte_vector & mask;
+
+        let hi = bx.context.new_rvalue_vector_perm(None, hi_nibble, hi_nibble, masked_lo);
+        let lo = bx.context.new_rvalue_vector_perm(None, lo_nibble, lo_nibble, masked_hi);
+
+        let shuffled = hi | lo;
+        let loaded = bx.context.new_bitcast(None, shuffled, v_type);
+        let elems: Vec<_> = (0..in_len)
+            .map(|i| {
+                let idx = bx.gcc_int(bx.u32_type, i as _);
```

Comment:
When it is known to be native types, `context.new_rvalue_from_int()` is preferred.

Reply:
Fixed in 0cc4c03.